core/src/main/java/io/questdb/griffin/SqlCompiler.java

```diff
@@ -1878,45 +1886,162 @@ private long copyOrderedStrTimestamp(TableWriter writer, RecordCursor cursor, Re
 
     private void copyTable(SqlExecutionContext executionContext, CopyModel model) throws SqlException {
         try {
+            int workerCount = executionContext.getWorkerCount();
+            ExpressionNode fileNameNode = model.getFileName();
+            final CharSequence fileName = fileNameNode != null ? GenericLexer.assertNoDots(GenericLexer.unquote(fileNameNode.token), fileNameNode.position) : null;
+            if (model.isParallel()) {
+                if (workerCount < 1) {
+                    throw SqlException.$(0, "Invalid worker count set [value=").put(workerCount).put("]");
+                }
+
+                MessageBus bus = executionContext.getMessageBus();
+                RingQueue<TextImportRequestTask> queue = bus.getTextImportRequestCollectingQueue();
+                if (queue.getCycle() < 1) {
+                    throw SqlException.$(0, "Parallel import requests queue size can't be zero!");
+                }
+                long correlationId = TextImportRequestTask.newTaskCorrelationId();
+                long timeout = configuration.getParallelImportRequestWaitTimeout();
+                if (model.isCancel()) {
+                    addTextImportRequest(correlationId, timeout, model, null);
+                } else {
+                    if (model.getTimestampFormat() == null) {
+                        model.setTimestampFormat("yyyy-MM-ddTHH:mm:ss.SSSUUUZ");
+                    }
+                    if (model.getDelimiter() < 0) {
+                        model.setDelimiter((byte) ',');
+                    }
+                    addTextImportRequest(correlationId, timeout, model, fileName);
+                }
+                awaitTextImportResponse(correlationId, timeout);
```

Comment:
What do we do when we only have a single shared worker? Any parallel import attempt would fail on such server since the only thread would be blocked by this await. Maybe we could fail fast in such scenario and return a meaningful SQL error suggesting the user to use non-parallel import?

Reply:
thread will be released on timeout

Reply:
so when i configure 1 shared worker and execute SQL `copy`:

```
2022-07-20T17:57:18.828930Z I i.q.c.TableWriter switched partition [path='C:\qdb\db\sys.parallel_text_import_log\2022-07-20.103']
2022-07-20T17:57:18.831185Z E i.q.c.t.TextImportRequestProcessingJob error saving to parallel import log table, unable to insert, releasing writer and stopping log table updates [table=sys.parallel_text_import_log, error=
java.lang.NullPointerException: Cannot invoke "io.questdb.cairo.MapWriter.put(java.lang.CharSequence)" because the return value of "io.questdb.std.ObjList.getQuick(int)" is null
	at io.questdb.cairo.TableWriter$RowImpl.putSym(TableWriter.java:5535)
	at io.questdb.cutlass.text.TextImportRequestProcessingJob.updateStatus(TextImportRequestProcessingJob.java:171)
	at io.questdb.cutlass.text.ParallelCsvFileImporter.updateStatus(ParallelCsvFileImporter.java:422)
	at io.questdb.cutlass.text.ParallelCsvFileImporter.phasePrologue(ParallelCsvFileImporter.java:1025)
	at io.questdb.cutlass.text.ParallelCsvFileImporter.parseStructure(ParallelCsvFileImporter.java:757)
	at io.questdb.cutlass.text.ParallelCsvFileImporter.process(ParallelCsvFileImporter.java:368)
	at io.questdb.cutlass.text.TextImportRequestProcessingJob.runSerially(TextImportRequestProcessingJob.java:147)
	at io.questdb.mp.SynchronizedJob.run(SynchronizedJob.java:39)
	at io.questdb.mp.Worker.run(Worker.java:116)
]
2022-07-20T17:57:18.831634Z I i.q.c.TableWriter tx rollback [name=sys.parallel_text_import_log]
2022-07-20T17:57:18.832001Z I i.q.c.TableWriter purging non attached partitions [path=C:\qdb\db\sys.parallel_text_import_log]
2022-07-20T17:57:18.832980Z I i.q.c.TableWriter switched partition [path='C:\qdb\db\sys.parallel_text_import_log\2022-07-18.49']

```

this is what i get in the console:

![image](https://user-images.githubusercontent.com/7276403/180050624-72a8b3e3-44ca-4461-936d-69812a99d190.png)

I can see three problems:

- import should run, to completion even on one thread, otherwise work stealing does not do what it is supposed to
- there should be no NPEs under any circumstances
- the timeout message should be rephrased to be easily understood to user (e.g. it should enable user to self-fix the issue)

Reply:
and another NPE on exit:

```
2022-07-20T18:10:55.043110Z E i.q.c.t.ParallelCsvFileImporter could not import [phase=PARTITION_IMPORT, ex=import cancelled [phase=PARTITION_IMPORT, msg=`Cancelled`]]
2022-07-20T18:10:55.043235Z C server-main unhandled error [job=io.questdb.cutlass.text.TextImportRequestProcessingJob@dc0829f, ex=
java.lang.NullPointerException: Cannot invoke "io.questdb.cairo.TableWriter.getPartitionCount()" because "this.writer" is null
	at io.questdb.cutlass.text.TextImportRequestProcessingJob.enforceLogRetention(TextImportRequestProcessingJob.java:116)
	at io.questdb.cutlass.text.TextImportRequestProcessingJob.runSerially(TextImportRequestProcessingJob.java:157)
	at io.questdb.mp.SynchronizedJob.run(SynchronizedJob.java:39)
	at io.questdb.mp.Worker.run(Worker.java:116)
]
2022-07-20T18:10:55.043450Z I http-server connected [ip=127.0.0.1, fd=1600]
2022-07-20T18:10:55.044401Z I i.q.c.h.HttpServer closed
```

Reply:
the first NPE is to do with my existing `log` table having old structure. Dropping this table fixed the NPE.

Reply:
with the table fixed, I now get `Parallel import ack timeout` but import is running in the background

Reply:
ok, to clarify the behaviour:

- there is no ack, because there is only one thread
- import still goes ahead, after ack timeout
- import finishes on single thread, no problem, work stealing is fine
- i would also add that import on single thread is relatively fast
- there are NPEs after i removed my local log table and let questdb create a new one
- I would perhaps rephrase the ack error, so user knows what's happening, will push a commit in a sec

otherwise all looks good

Reply:
The timeout is huge which is not nice. Moreover, we could detect the shared worker count == 1 situation and return immediately or at least use a smaller timeout.

Reply:
Forgot to mention an unpleasant side effect of running a lengthy import on a single worker pool - the database will stay unresponsive until the import finishes. Considering this and the fact that other SQL queries may lead to the same effect, I'd even propose limiting min shared worker count prop to 2, so that we don't get into such situations in simple scenarios.

As for the collecting queue, it feels like a potentially artificial construct that could be replaced with a synchronized method. The thread which handles a copy command can do the same thing as the collecting task does with the same effect.

Reply:
> could be replaced with a synchronized method

The method doesn't even have to be a synchronized one. Should be enough to change `textImportRequestProcessingPubSeq` to be a `MPSequence` and slightly tweak cancellation handling.

Reply:
> As for the collecting queue, it feels like a potentially artificial construct that could be replaced with a synchronized method. The thread which handles a copy command can do the same thing as the collecting task does with the same effect.

Synchronized method is irrelevant here.
The whole collection queue + job combination is artificial.
The real question is where to store the CancellationToken (well, CircuitBreaker) for the import task.
It should be globally accessible.

Reply:
> The real question is where to store the CancellationToken (well, CircuitBreaker) for the import task.

That's simple - `CairoEngine` to the rescue.

Reply:
But in this way, the `CairoEngine` will be turned into a trash can.
Actually, this is plan B after @bluestreak01 's approval.

Reply:
> * there is no ack, because there is only one thread
> * import still goes ahead, after ack timeout
It's unavoidable with current design. In any case, it's an ack timeout, not a request reject.

> * there are NPEs after i removed my local log table and let questdb create a new one
I cannot reproduce NPE

Reply:
NPE is a non-issue, I was talking to myself closer to the top of this thread, where I explained why. I feel you read half of what I wrote. I realise it was a lot to read though.

Reply:
> NPE is a non-issue, I was talking to myself closer to the top of this thread, where I explained why. I feel you read half of what I wrote. I realise it was a lot to read though.

I've read all your comments, but the summary statement claims:
there **ARE** NPEs after i removed my local log table and let questdb create a new one

Reply:
There are no NPEs anymore.

I had old log table, locally. It was causing issues until I removed it.